hal/src/thumbv7em/clock/v2/gclk.rs

```diff
@@ -0,0 +1,746 @@
+//! let (pclk_sercom0, gclk2) = Pclk::enable(tokens.pclks.sercom0, gclk2);
+//! ```
+
+use core::marker::PhantomData;
+
+use seq_macro::seq;
+use typenum::{U0, U1};
+
+use crate::pac;
+use crate::pac::NVMCTRL;
+
+pub use crate::pac::gclk::genctrl::{DIVSEL_A, SRC_A as GclkSourceEnum};
+
+use super::gclkio::NotGclkInput;
+use crate::clock::v2::{
+    types::{Counter, Decrement, Enabled, Increment, PrivateIncrement},
+    Source, SourceMarker,
+};
+use crate::pac::gclk::{RegisterBlock, GENCTRL};
+use crate::time::Hertz;
+use crate::typelevel::Sealed;
+
+//==============================================================================
+// GclkToken
+//==============================================================================
+
+/// A [`GclkToken`] equals a hardware register
+/// Provide a safe register interface for [`Gclk`]s
+///
+/// This `struct` takes ownership of a [`GclkNum`] and provides an API to
+/// access the corresponding registers
+pub struct GclkToken<G: GclkNum> {
+    gen: PhantomData<G>,
+}
+
+impl<G: GclkNum> GclkToken<G> {
+    /// Create a new instance of [`GclkToken`]
+    ///
+    /// # Safety
+    ///
+    /// Users must never create two simulatenous instances of this `struct` with
+    /// the same [`GclkNum`]
+    #[inline]
+    pub(super) unsafe fn new() -> Self {
+        GclkToken { gen: PhantomData }
+    }
+
+    /// Used to mask out the correct bit based on [`GclkNum`]
+    #[inline]
+    fn mask(&self) -> u16 {
+        1 << G::NUM
+    }
+
+    /// Provides the base pointer to the [`Gclk`] registers
+    ///
+    /// # Safety
+    ///
+    /// Only one [GclkToken] accessible at any given time
+    #[inline]
+    fn gclk(&self) -> &RegisterBlock {
+        unsafe { &*pac::GCLK::ptr() }
+    }
+
+    /// Provides a pointer to the individual Generator Control [`GENCTRL`]
+    /// registers
+    ///
+    /// Each GCLK 0 to 11 has its own Generator Control [`GENCTRL`] register
+    /// controlling the settings of that specific generator
+    #[inline]
+    fn genctrl(&self) -> &GENCTRL {
+        &self.gclk().genctrl[G::NUM]
+    }
+
+    /// Block until synchronization has completed
+    ///
+    /// Used for any registers annotated with
+    ///
+    /// * "Write-Synchronized"
+    /// * "Read-Synchronized"
+    ///
+    /// in the Property field
+    #[inline]
+    fn wait_syncbusy(&self) {
+        while self.gclk().syncbusy.read().genctrl().bits() & self.mask() != 0 {}
+    }
+
+    /// Set the clock source for the [`Gclk`] generator
+    #[inline]
+    fn set_source(&mut self, variant: GclkSourceEnum) {
+        self.genctrl().modify(|_, w| w.src().variant(variant));
+        self.wait_syncbusy();
+    }
+
+    /// When dividing an input clock with a odd division factor the duty-cycle
+    /// is not 50-50, enabling this ensures 50-50 duty-cycle on the
+    /// resulting generator clock
+    #[inline]
+    fn improve_duty_cycle(&mut self, flag: bool) {
+        self.genctrl().modify(|_, w| w.idc().bit(flag));
```

Comment:
@sakian, thanks for testing this out! Honestly, I didn't think anyone other than @vccggorski, @vcchtjader and myself would actually use it before merging.

It looks like this was already caught at some point, because the most recent code includes it.

For the moment, [my branch](https://github.com/bradleyharden/atsamd/tree/clocking-api-v2) is the most up-to-date version. We're reviewing my latest additions [here](https://github.com/vccggorski/atsamd/pull/5).

Sorry if we break your code, but this is still very much a WIP.

Reply:
Thanks, you're doing great work. Everything else worked flawlessly! I'll pull the latest and give it a test when I get a chance.

Reply:
By the way, if you dug into the code enough to find this bug, you probably have a reasonable understanding of the API. I think we would appreciate any feedback you have on the overall structure. Keep in mind, though, that the documentation is wildly out-of-date in some places. For the moment, the best source of knowledge is the code itself.